pkg/controller/installation/core_controller.go

```diff
@@ -1472,86 +1383,82 @@ func (r *ReconcileInstallation) SetDegraded(reason string, err error, log logr.L
+func GetOrCreateTyphaNodeTLSConfig(cli client.Client, certificateManager certificatemanagement.CertificateManager) (*render.TyphaNodeTLS, error) {
 	// accumulate all the error messages so all problems with the certs
 	// and CA are reported.
 	errMsgs := []string{}
-	ca, err := r.validateTyphaCAConfigMap()
+	configMap, err := validateTyphaCAConfigMap(cli)
 	if err != nil {
 		errMsgs = append(errMsgs, fmt.Sprintf("CA for Typha is invalid: %s", err))
 	}
-
-	node, err := utils.ValidateCertPair(
-		r.client,
-		common.OperatorNamespace(),
-		render.NodeTLSSecretName,
-		render.TLSSecretKeyName,
-		render.TLSSecretCertName,
-	)
-	if err != nil {
-		errMsgs = append(errMsgs, fmt.Sprintf("CertPair for Felix is invalid: %s", err))
-	} else if node != nil {
-		if node.Data != nil {
-			// We need the CommonName, URISAN, or both to be set
-			_, okCN := node.Data[render.CommonName]
-			_, okUS := node.Data[render.URISAN]
-			if !(okCN || okUS) {
-				errMsgs = append(errMsgs, fmt.Sprintf("CertPair for Felix does not contain common-name or uri-san"))
+	getKeyPair := func(secretName, commonName string) (keyPair certificatemanagement.KeyPair, cn string, uriSAN string) {
+		keyPair, err = certificateManager.GetKeyPair(cli, secretName, common.OperatorNamespace())
+		if keyPair == nil {
+			keyPair, err = certificateManager.GetOrCreateKeyPair(cli, secretName, common.OperatorNamespace(), []string{commonName})
+			if err != nil {
+				errMsgs = append(errMsgs, err.Error())
 			}
-		}
-	}
-
-	typha, err := utils.ValidateCertPair(
-		r.client,
-		common.OperatorNamespace(),
-		render.TyphaTLSSecretName,
-		render.TLSSecretKeyName,
-		render.TLSSecretCertName,
-	)
-	if err != nil {
-		errMsgs = append(errMsgs, fmt.Sprintf("CertPair for Typha is invalid: %s", err))
-	} else if typha != nil {
-		if typha.Data != nil {
-			// We need the CommonName, URISAN, or both to be set
-			_, okCN := typha.Data[render.CommonName]
-			_, okUS := typha.Data[render.URISAN]
-			if !(okCN || okUS) {
-				errMsgs = append(errMsgs, fmt.Sprintf("CertPair for Typha does not contain common-name or uri-san"))
+		} else if err != nil {
+			errMsgs = append(errMsgs, fmt.Sprintf("CertPair for %s is invalid: %v", secretName, err))
+		} else {
+			data := keyPair.Secret(common.OperatorNamespace()).Data
+			if data != nil {
+				cn, uriSAN = string(data[render.CommonName]), string(data[render.URISAN])
 			}
 		}
+		if !keyPair.BYO() {
+			cn = commonName
+		} else if cn == "" && uriSAN == "" {
+			errMsgs = append(errMsgs, fmt.Sprintf("CertPair for %s does not contain common-name or uri-san", secretName))
+		}
+		return
 	}
+	node, nodeCommonName, nodeURISAN := getKeyPair(render.NodeTLSSecretName, render.FelixCommonName)
+	typha, typhaCommonName, typhaURISAN := getKeyPair(render.TyphaTLSSecretName, render.TyphaCommonName)
 
 	// CA, typha, and node are all not set
-	allNil := (ca == nil && typha == nil && node == nil)
+	allOperatorProvided := configMap == nil && !typha.BYO() && !node.BYO()
```

Comment:
Why does this matter if it's BYO or not? This just cares if there are certificates or not (at least that's what it seems like reading what was previously there).

Reply:
This function has been refactored; it now combined getting the secrets and creating them if absent, whereas previously the creation was an entirely different function. For that reason, these lines changed: the keypairs are never nil at this point.